README.md

```diff
@@ -52,3 +52,25 @@ etc.). There is no implicit conversion to Michelson's binary format (`PACK`) nor
 quoting mechanism.
 
 After the conversion, the bytes can be use in the field `metadata` when originating the contract.
+
+## Error Code
+
+In error scenarios, the contract fails with an error code.
+
+Here is a summary of all the error codes thrown by the contract.
+(The list of errors may be inaccurate and incomplete, it will be updated during the implementation.)
+
+| Error Code       | Description                                                                                                              |
+|------------------|--------------------------------------------------------------------------------------------------------------------------|
+| 100              | Thrown by `get_tick`: Cannot get tick in tick indexed state, tick is not initialized.                                    |
+| 101              | Thrown by `initialize_tick`: Invalid withness.                                                                           |
+| 102              | Thrown by `collect_fees`: Position does not exist.                                                                       |
+| 103              | Thrown by `floor_log_half_bps`: Log out of bounds.                                                                       |
+| 104              | Thrown by `assert_nat`: X should be positive.                                                                            |
```

Comment:
@Martoon-00 So I decide to update `assert_nat` to accept another error code argument. Also added a list of all not nat errors in the table:

```
| 202              | Throw when `(p.dx - fee)` is not nat |
| 203              | Throw when `(p.s.sqrt_price - sqrt_price_new)` is not nat |
| 204              | Throw when `(dx_consummed - dx_for_dy)` is not nat |
| 205              | Throw when `(fee_growth_x_new - tick.fee_growth_outside.x)` is not nat |
| 206              | Throw when `(p.dx - dx_consummed)` is not nat. (TODO: similar to 202) |
| 207              | Throw when `(sqrt_price_new - p.s.sqrt_price)` is not nat |
| 208              | Throw when `(liquidity - tick.delta_liquidity)` is not nat |
| 209              | Throw when `(p.dx - r.dx)` is not nat |
...
```

Currently the message is really useful, but this set it up for a better improvement now or later on.

Also I add `errors.mligo` file contains all the error constants, and uses this instead of using a nat value directly.

```
...
(* Assertion error, tick should be initialized  *)
let internal_tick_not_initialized_err = 200n

(* Time now is smaller than epoch time. *)
let internal_epoch_bigger_than_now_err = 201n

(* Throw when `(p.dx - fee)` is not nat *)
let internal_202 = 202n
....
```

This way, instead of talking about error 200, we can use the label instead `internal_tick_not_initialized_err`, which is less confusing and the error code may change later on.

This file and `error-code.md` file are updated via a script `generate_error_code.hs` script.

Let me know what you think about this.

Reply:
Also I put error label for these various assert_nat as: `internal_203, internal_204 ...` as placeholders.

Let me know as well if you have suggestion on how to name these errors.

Reply:
I believe that it is possible to specify the semantics beyond each of the internal error (now or later). For some of them, it is quite easy to prove that these errors are not possible. Other error scenarios rely on the internal contract's invariants that we can probably mention in the error description.

* `202` - this apparently means that fees percent `const_fee_bps` is initialized to be higher than 10000 (100%).
* `203` - we can make `move_sqrt_price` return both the new `sqrt_price` and diff (as `nat`). It's internal property is that the price decreases, and the error that corresponds to violation of this property can be named respectively ("`move_sqrt_price` blowed up").
* `204` - this looks super impossible, like `a * 10000 / (10000 - fee_bps)` is obviously less than `a`, not even sure how to name it :/
* `205` - we can say that "flip for fee_growth_outside` failed, this error is about contract's inner invariants.
*  `208` - this is about tracking the liquidity at the current price, so something like "liquidity went blow zero" looks just fine.
* `210` - insufficient token balance?
* `211` and `212` - they are good, but I think it's also worth including under which `if` we are (`if s.i_c >= key.hi.i` or `if s.i_c >= key.lo.i`) because it is likely that one branch is correct and another is not.
* `213` - "number of positions underflow"?
* `219` - `sqrt_price` happened not to grow monotonically with tick indices (this is an invariant of the contract)
* `218` - here the reason is the same as in `219`, though I think these are still worth carrying in different error codes as you do now since it is under a dedicated condition.

Other errors are far less obvious, seem fine to me to leave as-is.

Reply:
I've updated most of these error, but I leave 203 as it is now, since there is a case where we want the diff to be negative.

I also make some changes to the error code (remove 204, and re-order the codes), there maybe more changes to the error code, so I think it would be better to refer to an error with its label instead (except for the internal_2##, since we haven't comeup with a name yet). 

I've made most of the changes, this should be good for re-review.

Reply:
Some problem with my commit not getting through, this is fixed now.

Reply:
Looks good, thanks!